### PR TITLE
NEW Implement drag and drop for elements

### LINKED ISSUE #285: Implement React based drag and drop for reordering elements in the editor
In elemental <= 3.x the GridField sortable component is used to allow reordering elements.

In the React equivalent editor (not using GridField components) we'll need to add a React component to allow this.

### Acceptance Criteria
- [x] Styling as per one of design directions (pending decision): https://projects.invisionapp.com/dsm/silver-stripe/silver-stripe/asset/components/5b8c6cbfc5ffe300176d8e22
- [x] Blocks can be dragged up or down the hierarchy of the blocks on a page, both collapsed or expanded blocks can be dragged. 
- [x] On hover content blocks have a ‘drag handle’ icon that is visible for mouse operated UI’s which can be used to click and drag, as well as showing an icon on hover, there is a shadow below the block as per designs.
- [x] Dragging can be activated anywhere on the block where there isn’t another action and the block is pressed for x time (use smart defaults from drag-drop library). If dragging is in process (On “click-and-hold” or “touch-and-hold”) of a block, the drag handle and shadow beneath the block appears. 
- [x] I receive an indication from the interface as to where I can drop the block. Highlighted by an indicator line
- [ ] Re-ordered blocks will become “modified” if previously published, the block will require publishing for the effects of re-ordering to be visible on the live site.

### Options:

* https://github.com/react-dnd/react-dnd: already used in silverstripe/asset-admin
* https://github.com/atlassian/react-beautiful-dnd

The handling of the reordering could be done with a GraphQL mutation on ElementalArea, providing the list of element IDs in the desired new sort order after a drop is completed. This could be a GraphQL type like `ElementIDList`, but might be a little unstructured in terms of what GraphQL expects.

Another option might be to put the mutation directly into the element that is dragged and dropped, providing a mutation signature like this:

```graphql
mutation updateElementSortOrder($id:ID!, $before:ID, $after:ID) {
  setElementSortOrder(Input: {
    ID: $id
    MoveAfter: $after
    MoveBefore: $before
  }) {
    ID
  }
}
```

In this case the element would need to set its sort order to +1 or -1 of the MoveAfter/MoveBefore element, then set new sort orders for all elements in the list to respect the updates.

Just possible ideas at this stage.

### PR DESCRIPTION
**Pull request by ScopeyNZ**
This adds drag and drop functionality to the new elemental editor

Fixes #285

### CODE REVIEW

File: client/src/components/ElementEditor/Header.js [hunk 1]
```diff
@@ -16,6 +17,19 @@ class Header extends Component {
     };
   }
 
+  componentDidUpdate() {
+    if (this.state.tooltipOpen && this.props.disableTooltip) {
```
Comment (raissanorth) [inline client/src/components/ElementEditor/Header.js]: Perhaps we could destructure these props and state attributes for consistency's sake?!

**Comment by ScopeyNZ** [comment]
Updated with feedback & fixed Behat which were actually all valid failures 😥

File: client/src/components/ElementEditor/ElementEditor.js [hunk 2]
```diff
@@ -1,15 +1,52 @@
...
+    this.state = {
+      dragTargetElementId: null,
+      dragSpot: null,
+    };
+
+    this.handleDragOver = this.handleDragOver.bind(this);
+    this.handleDragEnd = this.handleDragEnd.bind(this);
+  }
+
+  handleDragOver(element = null, isOverTop = null) {
```
[the remark this answers is not in the document]
Comment (ScopeyNZ) [reply client/src/components/ElementEditor/ElementEditor.js]: Done :)

**Comment by raissanorth** [comment]
Tested it locally and it's awesome! Great job.

**Comment by ScopeyNZ** [comment]
Done 🙂

File: client/src/components/ElementActions/PublishAction.js [hunk 3]
```diff
@@ -9,21 +9,20 @@ import i18n from 'i18n';
  * Adds the elemental menu action to publish a draft/modified block
  */
 const PublishAction = (MenuComponent) => (props) => {
+  const { element, actions: { handlePublishBlock } } = props;
```
Comment (NightJar) [inline client/src/components/ElementActions/PublishAction.js]: I think `element` here could become `element: { id, Title: title, BlockSchema, Version, isLiveVersion }` and remove a lot of the other changes (effectively unnecessary repetition).
Same with the other `Actions` HoCs.

Comment (ScopeyNZ) [reply client/src/components/ElementActions/PublishAction.js]: I considered this but thought it was being sort of over-verbose:

`element: { ID: id, Title: title, BlockSchema: { type }, Version: version, IsLiveVersion: isLiveVersion }`

We've not de-structured `element` in some other places too.

I will change it though - it doesn't really bother me what we end up doing.

Comment (NightJar) [reply client/src/components/ElementActions/PublishAction.js]: Yeah, tbh I was of 2 minds about it. Having the prefix makes it obvious where the value comes from too.
But I think it's that we prefer destructuring where possible.

Comment (NightJar) [reply client/src/components/ElementActions/PublishAction.js]: On second thought, I don't care that strongly about this.

**Pull request merged**